Describe display formats and various fixes (BGR565->RGB565X, RGB888, SDL driver)

avolmat-st opened the pull request:
Running the display sample on STM32 in RGB888 color leads to having red rectangle and blue rectangle swapped.
(RGB565 and ARGB8888 are correct).

Within the sample, the RGB888 red 0x00FF0000u is written into memory as
```c
        for (size_t idx = 0; idx < buf_size; idx += 3) {
                *(buf + idx + 0) = color >> 16;
                *(buf + idx + 1) = color >> 8;
                *(buf + idx + 2) = color >> 0;
        }
```
That is address 0: R, address 1: G then address 2: B

Having checked within LVGL but also other formats such as Linux DRM via https://github.com/afrantzis/pixel-format-guide tools, RGB888 is represented in memory as:

```
python3 -m pfg describe DRM_FORMAT_RGB888
Format:               DRM_FORMAT_RGB888
Component data type:  UNORM
Described as:         Bytes in memory
Memory little-endian: 0                1                2
                      M              L M              L M              L
                      B₇B₆B₅B₄B₃B₂B₁B₀ G₇G₆G₅G₄G₃G₂G₁G₀ R₇R₆R₅R₄R₃R₂R₁R₀
Memory big-endian:    0                1                2
                      M              L M              L M              L
                      B₇B₆B₅B₄B₃B₂B₁B₀ G₇G₆G₅G₄G₃G₂G₁G₀ R₇R₆R₅R₄R₃R₂R₁R₀
```

and for ARGB8888

```
python3 -m pfg describe DRM_FORMAT_ARGB8888
Format:               DRM_FORMAT_ARGB8888
Component data type:  UNORM
Described as:         Bytes in memory
Memory little-endian: 0                1                2                3
                      M              L M              L M              L M              L
                      B₇B₆B₅B₄B₃B₂B₁B₀ G₇G₆G₅G₄G₃G₂G₁G₀ R₇R₆R₅R₄R₃R₂R₁R₀ A₇A₆A₅A₄A₃A₂A₁A₀
Memory big-endian:    0                1                2                3
                      M              L M              L M              L M              L
                      B₇B₆B₅B₄B₃B₂B₁B₀ G₇G₆G₅G₄G₃G₂G₁G₀ R₇R₆R₅R₄R₃R₂R₁R₀ A₇A₆A₅A₄A₃A₂A₁A₀
```

In the display sample, the ARGB8888 corresponds well to this format, however not RGB888.

I think this is only a display sample issue since LVGL based program works well in RGB888 so the issue is only the data prepared within the display sample.
I've tested this on STM32 / LTDC based platform and this work well. It would be also great if it could be tested on some other platforms supporting RGB888 on display.

This PR corrects the RGB888 data within the sample and also add a description within the display header to avoid such future misunderstanding of the data ordering.

Commit:
include: display: Add description for all formats

Add detailed description for all supported display formats.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: display: fix RGB888 red/blue values

The RGB888 colors generated leads to having
  byte0 - R
  byte1 - G
  byte2 - B
while the RGB888 is usually described as B - G - R
going from byte0 to byte2.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

ngphibang commented:
Tested on native sim and RT1170, both works correctly. And logically, I think it should be fixed like that to have consistency between RGB565, RGB888 and ARGB8888 where the R channel is in the highest byte.

However, in Linux kernel doc, the `V4L2_PIX_FMT_RGB24` fomat is contradictory with this PR, it seems the doc is wrong ?
https://www.kernel.org/doc/html/v4.9/media/uapi/v4l/pixfmt-packed-rgb.html

ngphibang reviewed (changes requested):
Need to be reconsidered as works wrongly on native sim

avolmat-st commented:
Let me update this PR with overall check of formats as well as SDL display mapping.

Commit:
display: fix SDL display format handling

Switch SDL buffers from SDL_PIXELFORMAT_ARGB8888 to
SDL_PIXELFORMAT_BGRA32. BGRA32 is platform endianness independent
description, defined as below:

	Byte 0   Byte 1   Byte 2   Byte 3
	7......0 15.....8 23....16 31....24
	Bbbbbbbb Gggggggg Rrrrrrrr Aaaaaaaa

The driver is then updated in order to ensure that all buffers,
given via a PIXEL_FORMAT_ described format are properly converted
into the SDL_PIXELFORMAT_BGRA32. At the same time endianness
conversion code is also added in order to avoid taking assumption
that the running platform is little-endian and ensure that data
are properly read as expected.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
display: rename current BGR_565 format into RGB_565X

The format currently expected by devices and sample display
application for BGR_565 is actually RGB_565 format with bytes
swapped (not B / R swapped). That is:

PIXEL_FORMAT_RGB_565:
         * @code{.unparsed}
         *   7......0 15.....8
         * | gggBbbbb RrrrrGgg | ...
         * @endcode

current PIXEL_FORMAT_BGR_565:
         * @code{.unparsed}
         *   7......0 15.....8
         * | RrrrrGgg gggBbbbb | ...
         * @endcode

This is explained in both st7796s display driver but also is
what is generated by the sample display application. As a video
format (ex: V4L2), such format is not mentioned as BGR (for
which R and B are swapped) but RGB_565X.

Within the whole Zephyr tree, rename the curremt BGR_565 format
into RGB_565X in order to emphasis that this is a byte swapped
format rather than a B/G component swapped format.

This also correct the description of the format in display.h
file, which wasn't correct based on what was being used by
display driver or sample display app.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: display: ensure data proper endianness buffer fill

Ensure proper endianness when data are written via 16 or 32bit write.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: display: add corner colors in L_8 mode

Draw different greyscale in all 4 corners in L_8 mode.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

faxe1008 reviewed (commented):
This also needs a entry in the migration guide.

josuah commented on boards/lilygo/twatch_s3/Kconfig.defconfig:
```diff
@@ -4,7 +4,7 @@
 
 choice ST7789V_PIXEL_FORMAT
-	default ST7789V_BGR565
+	default ST7789V_RGB565X
```
> adopting something like RGB565_BE instead

If some naming leads to bugs, it sounds like enough justification to rename it to avoid that to keep happening.

Bugs are present through a lot of implementations for this particular format, which shows that the original naming is not optimal.

A comment saying "this is the same as V4L2 format XYZ" will help connecting things together.

ngphibang commented on drivers/display/Kconfig.sdl:
```diff
@@ -34,8 +34,8 @@ choice SDL_DISPLAY_DEFAULT_PIXEL_FORMAT
 
-	config SDL_DISPLAY_DEFAULT_PIXEL_FORMAT_BGR_565
-		bool "BGR 565"
+	config SDL_DISPLAY_DEFAULT_PIXEL_FORMAT_RGB_565X
```
Here and line 34, should it be RGB565(X) without underscore to be coherent with the definition and other enums ?

ngphibang commented on include/zephyr/drivers/display.h:
```diff
@@ -62,16 +62,16 @@ enum display_pixel_format {
 	 */
 	PIXEL_FORMAT_RGB_565		= BIT(4),
```
Could we take the chance to modify also the definition to avoid ambiguity ? i.e., drop "For example, in little-endian machine:" as it makes people think that in BE machine the organization will be different ... while the correct definition specifies that the byte 0 (lowest address) is always `gggBbbbb` and the byte 1 is always `RrrrrGgg` despite of LE or BE machine. I think the definition in `video.h` should also be modified a bit but it is anyways less confused than here.

ngphibang commented on drivers/mipi_dsi/Kconfig.mcux:
```diff
@@ -23,7 +23,7 @@ config MIPI_DSI_MCUX_2L_SMARTDMA
 	help
 	  Use SMARTDMA. This accelerator will automatically
-	  convert RGB565 input data to BGR565 (little endian to big endian),
+	  convert RGB565 input data to RGB565X (little endian to big endian),
```
Could we drop `(little endian to big endian)` to avoid ambiguity ?

ngphibang reviewed (commented):
The renaming BGR565 to RGB565X is correct as it is actually a byte swap not channel swap. And this does not change the logic (just the renaming) so don't need to test on all platforms I think. BTW, the 1st commit title and body should be rewritten to be coherent with the naming (RGB565X not RGB565_BE).

The actual (logic) fix is only in the sdl display driver (last commit).

I retested anyways on native sim and RT1170, it works fine. Thanks.

ngphibang commented:
Even with the white background, it takes that long to fill the whole screen with white colors. Maybe because the eLCDIF driver allocates double buffers on the heap and does the memcpy (when the framebuffer size is not of the full display resolution 720p) ?

Commit:
samples: display: use white background for ARGB_8888

Set the background color to 0xFF (white) in case of ARGB_8888
framebuffer format so that it is same for all formats except
AL88 and MONO.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

ngphibang reviewed (dismissed):
LGTM. Thanks !

sylvioalves reviewed (changes requested):
I know that display driver is not stable yet, but there are changes in `display.h` that breaks current API.

At least I think migration notes update would be needed:
```
Display
=======

* The ``PIXEL_FORMAT_BGR_565`` pixel format has been renamed to
  :c:macro:`PIXEL_FORMAT_RGB_565X` to correctly reflect that it is a
  byte-swapped version of RGB_565, not a channel-swapped format.
  Applications using ``PIXEL_FORMAT_BGR_565`` must update to use
  ``PIXEL_FORMAT_RGB_565X``. (:github:`99276`)

* The devicetree macro ``PANEL_PIXEL_FORMAT_BGR_565`` has been renamed to
  ``PANEL_PIXEL_FORMAT_RGB_565X``. (:github:`99276`)

* The Kconfig options ``SDL_DISPLAY_DEFAULT_PIXEL_FORMAT_BGR_565`` and
  ``ST7789V_BGR565`` have been renamed to
  ``SDL_DISPLAY_DEFAULT_PIXEL_FORMAT_RGB_565X`` and ``ST7789V_RGB565X``
  respectively. (:github:`99276`)
```
